### PR TITLE
Windows support

### PR DESCRIPTION
**Pull request by peirick**

### CODE REVIEW

**Comment by aep** [comment]
lgtm, but confusing my osx broke. investigating.

File: ci.ps1 [hunk 1]
```diff
@@ -0,0 +1,25 @@
+$env:TARGET_CC="clang"
```
Comment (aep) [inline ci.ps1]: is this powershell? does every CI service have this? 
could you enable travis windows build or would you recommend something else?

File: modules/time/src/native.h [hunk 2]
```diff
@@ -1,10 +1,84 @@
 #include <stdint.h>
+#include <time.h>
 
 int os_time_tick(uint64_t *secs, uint64_t* nanos);
 int os_time_real(uint64_t *secs, uint64_t* nanos);
 
 #if defined(__linux__) || defined(__APPLE__)
-#include <time.h>
+#else
```
Comment (aep) [inline modules/time/src/native.h]: windows isnt not linux and not osx. it could be bare metal. this really needs a different guard here for windows

File: modules/time/src/native.h [hunk 3]
```diff
@@ -1,10 +1,84 @@
...
+#define NS_PER_US       1000ULL
+
+#define HNS_PER_SEC     (MS_PER_SEC * US_PER_MS * HNS_PER_US)
+#define NS_PER_HNS      (100ULL)    // NS = nanoseconds
+#define NS_PER_SEC      (MS_PER_SEC * US_PER_MS * NS_PER_US)
+
+#define CLOCK_MONOTONIC 1
+#define CLOCK_REALTIME  2
+
+int clock_gettime_monotonic(struct timespec *tv)
```
Comment (aep) [inline modules/time/src/native.h]: these should probably be static inline, otherwise the helpers get exposed into the resulting binary

File: src/make.rs [hunk 4]
```diff
@@ -173,7 +174,7 @@ impl Make {
             }
         }
 
-        cflags.push("-fPIC".into());
```
Comment (aep) [inline src/make.rs]: ah yes, here's how osx broke. this is wrong.

fpic is not a thing on windows?

File: modules/time/src/native.h (hunk 2 above)
Comment (peirick) [reply modules/time/src/native.h]: I had a look into this. #if defined(_WIN32) does work.

File: modules/time/src/native.h (hunk 3 above)
Comment (peirick) [reply modules/time/src/native.h]: yes. i'll change that.

File: ci.ps1 (hunk 1 above)
Comment (peirick) [reply ci.ps1]: following https://docs.travis-ci.com/user/reference/windows/#supported-languages powershell or even better bash or shell should work. I'll try to enable travis windows build

File: src/make.rs (hunk 4 above)
Comment (peirick) [reply src/make.rs]: The PIC flag does not work on windows. 

clang reports the following error:
> langclangclang: : errorclangerror: : : : error: error: unsupported option '-fPIC' for target 'x86_64-pc-windows-msvc'
unsupported option '-fPIC' for target 'x86_64-pc-windows-msvc'unsupported option '-fPIC' for target 'x86_64-pc-windows-msvc'unsupported option '-fPIC' for target 'x86_64-pc-windows-msvc'

I'll try to implement it similar to `asan`.

**Comment by aep** [comment]
this is looking great!

if passing the tests is too much work, i would suggest to enable travis in a new PR so we at least can get basic windows support in

File: modules/time/src/native.h [hunk 5]
```diff
@@ -1,10 +1,83 @@
 #include <stdint.h>
+#include <time.h>
 
 int os_time_tick(uint64_t *secs, uint64_t* nanos);
 int os_time_real(uint64_t *secs, uint64_t* nanos);
 
-#if defined(__linux__) || defined(__APPLE__)
```
Comment (aep) [inline modules/time/src/native.h]: why did you remove this? now the posix code is enabled on non-posix as well

Comment (peirick) [reply modules/time/src/native.h]: opsi. that was not intentional. I will correct this.

**Comment by peirick** [comment]
I got some problems that randomly some moduls like err, string or slice are not found. This seems to be only a problem on windows. Having a look at the logs on travis  this seems to be a bug with the loader.rs. Do you have any idea how to fix this ?

**Comment by aep** [comment]
my best guess would be that something is wrong here https://github.com/aep/zz/blob/master/src/lib.rs#L90

this is the default search paths and in this case these modules are in CARGO_MANIFEST_DIR/modules

**Comment by peirick** [comment]
Got that bug. Would have been some combination of set_current dir and relative filepaths. Btw. Forward slashes are no problem on windows.

Next Bug is something with __attribute__ (visibility

**Comment by aep** [comment]
Great work! Thank you so much.

Unfortunately in a family emergency and don't have much attention span right now so just a quick note 

Visibility is a big part of exporting a clean abi. 

"Pub" functions need to be removed from the abi and they can't be static because other modules need them. 

Windows does have an equivalent thing, I remember from Qt.

**Comment by aep** [comment]
emitter.rs should not emit an attribute but instead some generic macro that is expanded in some sort of prelude.h

That makes it easier to later have it overwritable by the user.

**Comment by peirick** [comment]
thank you. take your time. familly is important. 

that macro would be something like mentioned here: https://stackoverflow.com/questions/19418908/creating-a-portable-library-to-run-on-both-linux-and-windows
correct? that prelude.h would be somewhere in the std module?

**Comment by aep** [comment]
dllexport/import looks correct to me yes.

i'd simply emit the boilerplate at the start of every c file for now. we can make that pretty later.
https://github.com/aep/zz/blob/master/src/emitter.rs#L64

**Comment by aep** [comment]
amazing!
thank you again, that's some great work
